traits/trait_type.py

```diff
@@ -305,16 +330,46 @@ def clone(self, default_value=NoDefaultSpecified, **metadata):
         new._metadata.update(metadata)
 
         if default_value is not NoDefaultSpecified:
-            if self.validate is not None:
+            if new.default_value_type in clone_no_override_default_value:
+                raise TraitError(
+                    "Can't override default value of cloned {new} trait",
+                )
+
+            # does the trait want the original value as the default?
+            # TODO: since this is needed here, it should also be a property of
+            # the TraitType, not just the CTrait.
+            setattr_original_value = new.as_ctrait().setattr_original_value
+
+            # try to validate the new default value
+            if not (self.validate is None or setattr_original_value):
                 try:
                     default_value = self.validate(None, None, default_value)
-                except Exception:
+                except Exception as exc:
+                    # this is expected for traits which need object or name
+                    # information to properly validate (eg. This() trait)
+                    # TODO: override This.clone() and similar for better
+                    # behaviour and raise a TraitError here
+                    logger.debug(
+                        "Validation failed cloning %s with default value %s: %s",  # noqa: E501
+                        self,
+                        default_value,
+                        exc,
+                    )
                     pass
 
-            # Known issue: this doesn't do the right thing for
-            # List, Dict and Set, where we really want to make a copy.
-            # xref: enthought/traits#1630
-            new.default_value_type = DefaultValue.constant
+            if new.default_value_type in clone_copies_default_value:
+                default_value = default_value.copy()
+
+            if new.default_value_type in clone_becomes_constant_default_value:
+                # Known issue: this doesn't do the right thing if the default
```

Comment:
I'd argue that this _is_ the right thing at this level, because `TraitType` itself has no way to know about the needs of particular `TraitType` subclasses. And in many cases the value might be mutable but it _still_ might be the right thing to share it between instances (e.g., if it's technically mutable but not intended to be used mutably). So yes, we should document the pattern that `TraitType` subclasses should override / extend the appropriate methods if they want to do something more sophisticated.

Reply:
I've reworded slightly in a328c639e457349bb92a95e2f64a285be48bbbe0.